addons/account/models/account_move_line.py

```diff
@@ -1847,94 +1840,483 @@ def get_accounting_rate(vals):
+        for index, exchange_move in zip(exchange_diff_partial_index, exchange_moves):
+            partials[index].exchange_move_id = exchange_move
+
+        # ==== Create entries for cash basis taxes ====
+        def is_cash_basis_needed(account):
+            return account.company_id.tax_exigibility \
+                and account.account_type in ('asset_receivable', 'liability_payable')
+
+        if not self._context.get('move_reverse_cancel'):
+            for plan in plan_list:
+                if is_cash_basis_needed(plan['amls'].account_id):
+                    plan['partials']._create_tax_cash_basis_moves()
+
+        # ==== Prepare full reconcile creation ====
+        # First, we need to find all sub-set of amls that are candidates for a full.
+
+        def is_line_reconciled(aml, has_multiple_currencies):
+            # Check if the journal item passed as parameter is now fully reconciled.
+            if aml.reconciled:
+                return True
+            if not aml.matched_debit_ids and not aml.matched_credit_ids:
+                # Suppose a journal item having balance = 0 but an amount_currency like an exchange difference.
+                return False
+            if has_multiple_currencies:
+                return aml.company_currency_id.is_zero(aml.amount_residual)
+            else:
+                return aml.currency_id.is_zero(aml.amount_residual_currency)
+
+        full_batches = []
+        all_aml_ids = set()
+        for plan in plan_list:
+            for aml in plan['amls']:
+                if 'full_batch_index' in aml_values_map[aml]:
+                    continue
+
+                involved_amls = plan['amls']._all_reconciled_lines()
+                all_aml_ids.update(involved_amls.ids)
+                full_batch_index = len(full_batches)
+                has_multiple_currencies = len(involved_amls.currency_id) > 1
+                is_fully_reconciled = all(
+                    is_line_reconciled(aml, has_multiple_currencies)
+                    for aml in involved_amls
+                )
+                full_batches.append({
+                    'amls': involved_amls,
+                    'is_fully_reconciled': is_fully_reconciled,
+                })
+                for aml in involved_amls:
+                    if aml_values_map.get(aml):
+                        aml_values_map[aml]['full_batch_index'] = full_batch_index
+
+        # ==== Prefetch the fields all at once to speedup the reconciliation ====
+        # All of those fields will be cached by the orm. Since the amls are split into multiple batches, the orm is not
+        # able to prefetch the data for all of them at once. For that reason, we force the orm to populate the cache
+        # before doing anything.
+        all_amls = self.browse(list(all_aml_ids))
+        all_amls.move_id
+        all_amls.matched_debit_ids
+        all_amls.matched_credit_ids
```

Comment:
Didn't we already do that at line 2092 ?
I see that we browse `all_aml_ids` but didn't `_optimize_reconciliation_plan` already does the gathering once?
I feel like we are doing the exact same thing twice but I may be missing something.

Reply:
The problem is we create the partials/full at this point so we need to redo the pre-fetching since the data have been invalidated

Reply:
It makes sense indeed. You should explain it above too then so it is crystal clear